Support for the startDebugging reverse request

jasonwilliams opened the pull request:
This PR adds support for the [startDebugging reverse request](https://microsoft.github.io/debug-adapter-protocol/specification.html#Reverse_Requests_StartDebugging). It achieves this by adding support for child debugger which is essentially the same client as the parent (copies the same config), but can send different launch values. The child debugger is then added to the slotmap which holds the debuggers.

Fixes: https://github.com/helix-editor/helix/issues/11906
Fixes: https://github.com/helix-editor/helix/issues/4214

The startDebugging request is used by the JavaScript debugger and I suspect future debug adapters will use it when new threads are discovered too. I have got this working and the breakpoints hit, but some tidy up is needed

![Screenshot From 2025-05-11 18-04-47](https://github.com/user-attachments/assets/9d88e967-7fc5-459e-a5d3-dcff9322bae2)

This PR also opens the door for having 2 top-level debuggers run simultaneously (which you can do in other editors), although right now that's still disabled as it would still be a bit of work.


## How to test

- echo `console.debug("hello world")' > /tmp/index.js`
- wget 'https://github.com/microsoft/vscode-js-debug/releases/download/v1.97.1/js-debug-dap-v1.97.1.tar.gz' -O /tmp/js-debug-dap.tar.gz
- tar -xvzf /tmp/js-debug-dap.tar.gz
- Apply [languages.toml](https://github.com/helix-editor/helix/issues/11906#issue-2593059061) configuration to ~/.config/helix/languages.toml
- `hx -v /tmp/index.js`
- `<space>Gbl<return>index.ts<return><esc>`

## Changes
This not only adds support for multiple debuggers by removing the assumption there is only  1 debug client and instead holding a hashmap of clients which may be active. This is [inline with the specification](https://microsoft.github.io/debug-adapter-protocol//specification.html#Reverse_Requests_StartDebugging) and with how other IDEs like VSCode treat them too. This should make life easier for ownership, as removing 1 debugger will also remove its children automatically.

### Logging change
Logs have been updated to show which debug adapter is making the request or receiving the response. You now have the adapter ID before the direction of traffic in the form of `[<ID>] <- DAP <Data>`. Here is an example showing 2 debug adapters being used during a js handshake

```
2025-05-05T17:03:52.412 helix_dap::transport [INFO] [2] <- DAP success in response to 0
2025-05-05T17:03:52.412 helix_dap::transport [INFO] [2] -> DAP {"type":"request","seq":1,"command":"launch","arguments":{"__pendingTargetId":"447da55cf39ed22c1eac5473","name":"index.js [128777]","type":"pwa-node"}}
2025-05-05T17:03:52.412 helix_dap::transport [INFO] [2] <- DAP {"seq":2,"type":"event","event":"initialized","body":{}}
2025-05-05T17:03:52.413 helix_dap::transport [INFO] [2] <- DAP event Event { event: "initialized", body: Some(Object {}) }
2025-05-05T17:03:56.850 helix_dap::transport [INFO] [0] <- DAP {"seq":9,"type":"event","event":"output","body":{"category":"telemetry","output":"js-debug/dap/operation","data":{"errors":[],"launch":{"operation":"launch","totalTime":62.1,"max":62.1,"avg":62.1,"stddev":null,"count":1,"failed":0},"!launch.errors":[],"launch.errors":[]}}}
2025-05-05T17:03:56.851 helix_dap::transport [INFO] [0] <- DAP event Event { event: "output", body: Some(Object {"category": String("telemetry"), "data": Object {"!launch.errors": Array [], "errors": Array [], "launch": Object {"avg": Number(62.1), "count": Number(1), "failed": Number(0), "max": Number(62.1), "operation": String("launch"), "stddev": Null, "totalTime": Number(62.1)}, "launch.errors": Array []}, "output": String("js-debug/dap/operation")}) }
2025-05-05T17:03:56.898 helix_dap::transport [INFO] [0] <- DAP {"seq":10,"type":"event","event":"output","body":{"category":"telemetry","output":"js-debug/cdp/operation","data":{"errors":[],"Target.targetCreated":{"operation":"Target.targetCreated","totalTime":0.1,"max":0.1,"avg":0.1,"stddev":null,"count":1,"failed":0},"!Target.targetCreated.errors":[],"Target.targetCreated.errors":[]}}}
2025-05-05T17:03:56.898 helix_dap::transport [INFO] [0] <- DAP event Event { event: "output", body: Some(Object {"category": String("telemetry"), "data": Object {"!Target.targetCreated.errors": Array [], "Target.targetCreated": Object {"avg": Number(0.1), "count": Number(1), "failed": Number(0), "max": Number(0.1), "operation": String("Target.targetCreated"), "stddev": Null, "totalTime": Number(0.1)}, "Target.targetCreated.errors": Array [], "errors": Array []}, "output": String("js-debug/cdp/operation")}) }
```

### Debugger updates
`startDebugging` is now supported and the client will attempt to connect a second debugger to the same adapter which is needed for debug adapters such as the JS debug adapter.
In order to make this happen we need to introduce the concept of a child debugger, which is just another client spawned by the parent and owned by the parent debugger. This allows multiple threads to be debugged at the same time and cleaned up properly when  a session ends.

DAP terminate was calling  'disconnect' instead of 'terminate', I've fixed this in this PR and the correct command is now being called under the hood. 

A bug has been fixed where the "Terminated" event was triggering the debug client to restart a session. This was because the client was only checking for the presence of "restart" but not what it's value was set to, this checks for both restart and it's value being set to `true`.

### Languages File
The languages files has been updated to point to the proper debug adapter server [location](https://github.com/microsoft/vscode-js-debug/releases) for javascript and not the deprecated version which was there before.

## Todo
- [x] Dap.terminate should use the "terminate" command, not "disconnect".
- [x] Fix bug where terminated event triggers restart due to the presence of the restart Boolean 
- [x] The child debugger should share the same properties as the parent. 
- [x]  skip `<node_internals>/**`
- [x] ~~Wrap debug commands with the debugger service so the IDs are set properly~~ pass the new ID from the service
- [x] On the thread event, set that thread to the active thread
- [x] Make sure debugging functionality works
- [x] Test also with Clang Debugger
- [x] Make connection type de-serializable

## Future updates or Questions
- Turn off debugger dialogue when hitting "t" for termintate
- Show active debuggers running in a picker view
- As it can take a while for the debuggers to connect (potential bug I think), dap should also share the spinner that LSP uses

jasonwilliams commented on helix-dap/src/client.rs:
```diff
@@ -323,6 +328,47 @@ impl Client {
+        };
+
+        // Use the same config as the parent to start a new child debugger
+        let result = Client::process(
```
For now this is hardcoded to create a new process, but some DAP servers support multiplexing clients to the same server, in future we may want to use TCP if the server sends an `attach` request.

the-mikedavis commented on helix-dap/src/debuger_service.rs:
```diff
@@ -0,0 +1,104 @@
+/// DebuggerService is a struct that manages and owns multiple debugger clients
+/// This holds the responsibility of managing the lifecycle of each client
+/// plus showing the heirarcihical nature betweeen them
+pub struct DebuggerService {
+    /// The top-level debuggers being held by this service
+    debuggers: HashMap<usize, Client>,
+    /// The active debugger client
+    ///
+    /// TODO: You can have multiple active debuggers, so the concept of a single active debugger
+    /// may need to be changed
+    current_debugger_id: Option<usize>,
+    /// This is used to generate unique ids for each debugger
+    pub counter: usize,
+}
```
See the `LanguageServerId` type and the use of `HopSlotMap` for language servers - that can be used instead of the hashmap+counter here and we probably don't need a new struct for this

jasonwilliams replied:
Thanks for looking.

I don’t know if HopSlotMap will work as you need to know the ID when creating the Client, so it’s too late by the time it’s added to the map. I believe this is different with language servers

the-mikedavis replied:
The LSP registry uses [`HopSlotMap::try_insert_with_key`](https://docs.rs/slotmap/latest/slotmap/struct.HopSlotMap.html#method.try_insert_with_key). It should be possible to use that for DAP as well to get the next ID and pass it to `Client::tcp`/`Client::process`.

jasonwilliams replied:
This has been resolved in the latest update

the-mikedavis commented on helix-dap/src/client.rs:
```diff
@@ -33,31 +33,32 @@ pub struct Client {
     pub caps: Option<DebuggerCapabilities>,
     // thread_id -> frames
     pub stack_frames: HashMap<ThreadId, Vec<StackFrame>>,
     pub thread_states: ThreadStates,
-    pub thread_id: Option<ThreadId>,
+    /// The currently active thread (usually the one we've broken at).
+    pub active_thread_id: Option<ThreadId>,
```
Please avoid renaming things, it only makes the diff harder to follow. Adding docs explaining its use is good though

jasonwilliams replied:
Sure, I can revert it. This was to be inline with `active_frame` just below

the-mikedavis commented on helix-dap/src/types.rs:
```diff
@@ -840,7 +870,7 @@ pub mod events {
     pub struct TerminatedBody {
         #[serde(skip_serializing_if = "Option::is_none")]
-        pub restart: Option<Value>,
+        pub restart: Option<bool>,
```
This can be an `any` according to the spec so it should stay as `Value` https://microsoft.github.io/debug-adapter-protocol//specification.html#Events_Terminated

It sounds like we shouldn't be interpreting this argument and instead should be passing it as-is to `Client::launch` or `Client::attach` if it is `Value::Bool(true)` or `Value::Object(_)`

jasonwilliams replied:
Yes you’re right, I will change this

jasonwilliams replied:
I think i was confusing this with disconnect's restart argument which does need to be a boolean. This has been fixed now.

the-mikedavis commented on helix-dap/src/types.rs:
```diff
@@ -992,6 +1022,29 @@ pub mod events {
     }
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
+pub enum ConnectionType {
```
Instead of manually implementing Deserialize this it should work to `#[derive(Deserialize)]` along with `#[serde(rename_all = "lowercase")]`:

```suggestion
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ConnectionType {
```

jasonwilliams commented:
In my latest changes...

- DAP now has a registry which operates very similar to the LSP registry.
- Moved some deps to top level so both dap and lsp can share the same versions.
- Moved some `client_start` logic into the registry so it can be reused by `startDebugging`
- Renamed `debuggers` to `dap_servers` to align with `lsp_servers`

the-mikedavis commented on helix-dap/src/registry.rs:
```diff
@@ -0,0 +1,114 @@
+}
+
+slotmap::new_key_type! {
+    pub struct DebugAdapterID;
```
```suggestion
    pub struct DebugAdapterId;
```

to match `LanguageServerId`

the-mikedavis commented on helix-view/src/editor.rs:
```diff
@@ -1083,8 +1082,7 @@ pub struct Editor {
 
-    pub debugger: Option<dap::Client>,
-    pub debugger_events: SelectAll<UnboundedReceiverStream<dap::Payload>>,
+    pub dap_servers: dap::registry::Registry,
```
```suggestion
    pub debug_adapters: dap::registry::Registry,
```

to match `language_servers`

the-mikedavis commented on helix-view/src/handlers/dap.rs:
```diff
@@ -393,10 +438,66 @@ impl Editor {
+
+                        let result = self
+                            .dap_servers
+                            .start_client(Some(socket), &config.expect("No config found"));
```
Instead of an `expect` let's `log::error!` when the config isn't set and `return true`

the-mikedavis commented on helix-lsp/Cargo.toml:
```diff
@@ -19,14 +19,14 @@ helix-loader = { path = "../helix-loader" }
 serde_json = "1.0"
 tokio = { version = "1.45", features = ["rt", "rt-multi-thread", "io-util", "io-std", "time", "process", "macros", "fs", "parking_lot", "sync"] }
-tokio-stream = "0.1.17"
+tokio-stream.workspace = true  
```
```suggestion
tokio-stream.workspace = true
```

looks like some trailing whitespace snuck in here

the-mikedavis commented on helix-dap/src/lib.rs:
```diff
@@ -1,8 +1,10 @@
 
-pub use client::{Client, ConnectionType};
+pub use client::Client;
+pub use futures_executor::block_on;
```
Looks like this is unused

the-mikedavis commented on helix-dap/Cargo.toml:
```diff
@@ -22,6 +22,11 @@ serde = { version = "1.0", features = ["derive"] }
+slotmap.workspace = true
+futures-executor.workspace = true
+futures-util.workspace = true
+tokio-stream.workspace = true  
```
```suggestion
tokio-stream.workspace = true
```

the-mikedavis reviewed (commented):
This looks good :+1:. I have a few minor comments
